### PR TITLE
ResponseEmitter added

### PR DESCRIPTION
**Pull request by plashenkov**

### CODE REVIEW

**Comment by Nyholm** [comment]
I’ve been meaning to create such PR myself. Could I politely suggest to put this class in a separate package? There is (nor will it be) anything guzzle related here. Right?

**Comment by plashenkov** [comment]
Let's discuss details here?
https://github.com/guzzle/psr7/issues/204

**Comment by plashenkov** [comment]
It seems so native to me to have this functionality right here, out of the box. We can read server request (::fromGlobals), we should be able to send the response as well, i.e. in and out.

**Comment by Nyholm** [comment]
Just FYI: There is a package doing just this: 
https://github.com/zendframework/zend-httphandlerrunner

**Comment by tflori** [comment]
@Nyholm in fact there are several packages. some are very simple packages and some are more complex. for sure this is also a very simple task. I can add it to every project I'm creating as well.

IMHO it should be a helper function `send(MessageInterface)`. There is a function `str(MessageInterface)` that is doing almost the same and has nothing to do with guzzle/psr-7 (but indeed with psr-7 itself).

**Comment by tflori** [comment]
I'm wondering about this check for output buffer. there are two concerns:

1. the reason for output buffering is to be able send headers after the output started (for example from a bad written code)
2. you can avoid this exception by opening another output buffer (an empty level).

I also have two concerns about the `emitHeaders()` method:

1. the headers should not be normalized - you are assuming here that every client is aware of the default form of headers (i. e. `Content-Type` not `content-type`) and you don't know which client requested the response the developer using this library will know.
2. As stated out in  [RFC 7230]() 

> A sender MUST NOT generate multiple header fields with the same field
   name in a message unless either the entire field value for that
   header field is defined as a comma-separated list [i.e., #(values)]
   or the header field is a well-known exception (as noted below).

Therefore in my opinion the exceptions should be hard coded here and other fields should just be defined as comma separated list. And I'm aware that a developer using this library could also think of this before he uses `withAddedHeader()`.

**Comment by plashenkov** [comment]
> I'm wondering about this check for output buffer.
Sorry, which check?

> the headers should not be normalized
Headers are not normalized by default. They will be normalized only if you want to turn it on (see constructor param). Compare with Zend Response Emitter: https://github.com/zendframework/zend-httphandlerrunner/blob/c865dbc17524a27285d99d1638afb3c5200a2035/src/Emitter/SapiEmitterTrait.php#L81 Here normalization is hardcoded and you have no way to turn it off. My opinion was that we should not touch headers at all, leave them to the developer. But other libraries include normalization (often hardcoded), this is why I added it too. But it is optional and turned off by default.

> Therefore in my opinion the exceptions should be hard coded here
If I understand you correctly... Well, I did not see other implementations use this approach. It would be nice to hear others' opinions. It seems to me that we should emit headers as they are defined by a developer. For some cases such custom helpful behaviour is not the thing they expect.

**Comment by tflori** [comment]
> Sorry, which check?

The check for output buffering in `assertNoPreviousOutput()`.

> Headers are not normalized by default.

You, are right. I missed the if around.

> It would be nice to hear others' opinions.

I agree that it makes sense to hear how others think the emitter should be have for multiple headers.

File: src/ResponseEmitter.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
+<?php
+namespace GuzzleHttp\Psr7;
+
+use Psr\Http\Message\ResponseInterface;
+
+/**
+ * Response emitter
+ */
+class ResponseEmitter
```
Comment (gmponos) [inline src/ResponseEmitter.php]: how about creating an Interface for this?

File: src/ResponseEmitter.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        }
+    }
+
+    /**
+     * Normalizes header name. For example: Content-type -> Content-Type.
+     *
+     * @param string $name
+     * @return string
+     */
+    protected function normalizeHeader($name)
```
Comment (gmponos) [inline src/ResponseEmitter.php]: Can this function be replaced by this? https://github.com/guzzle/psr7/blob/master/src/functions.php#L162

File: src/ResponseEmitter.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+ * Response emitter
+ */
+class ResponseEmitter
+{
+    /**
+     * Emits a response.
+     *
+     * @param ResponseInterface $response A response to emit.
+     * @param int $maxBufferSize Maximum output buffering size (pass 0 to emit all content at once).
+     * @param bool $normalizeHeaders Normalize header names while emitting. For example: Content-type -> Content-Type.
```
Comment (gmponos) [inline src/ResponseEmitter.php]: Checking other files comments should be alinged.. same for other functions that exist here.

File: src/ResponseEmitter.php (hunk 2 above)
Comment (plashenkov) [reply src/ResponseEmitter.php]: Nope. These functions are different despite similar names (see PHPDoc description).

File: src/ResponseEmitter.php (hunk 3 above)
Comment (plashenkov) [reply src/ResponseEmitter.php]: Yes, it looks like this is a standard here. This is a matter of dispute, strictly speaking. For example, here: https://github.com/guzzle/psr7/blob/master/src/Uri.php#L206-L209 and here: https://github.com/guzzle/psr7/blob/master/src/Uri.php#L247-L250 This is a code, not a PHPDoc, but the logic is the same. Somebody love to align all `===`, etc. When renaming any of variables or function names we need to realign everything every time. If we need to align, I'll do this for sure.

P.S. https://github.com/guzzle/psr7/blob/master/src/PumpStream.php#L34-L41 This block needs fix :)

File: src/ResponseEmitter.php (hunk 1 above)
Comment (plashenkov) [reply src/ResponseEmitter.php]: Maybe. Let's decide all together?

File: src/ResponseEmitter.php (hunk 3 above)
Comment (gmponos) [reply src/ResponseEmitter.php]: Yes.. guzzle repos do not have codestyles applied through their CI.. which makes it easier to accept different codestyles... 

I have created a [PR](https://github.com/guzzle/guzzle/pull/2194) for codestyle but it is [pending](https://github.com/guzzle/guzzle/pull/2167#issuecomment-438008642)

**Comment by sagikazarmark** [comment]
How about we just suggest zendframework/zend-httphandlerrunner? I really would like to avoid duplicating efforts, especially because this does not seem to be as trivial as it sounded before.

**Comment by sagikazarmark** [comment]
> without the need to plug another packages from another universes

I don't think that should be an issue.

Also, this PR seems to be a direct copy of the Zend SAPI Emitter, so if that code works, why can't we just recommend to install it. Originally that package was part of Diactoros, but I guess it was extracted for a reason.

Guzzle PSR-7 is mostly targeted at client side use cases, so if you need to use it on the server side, chances are you will have other server components installed anyway.

**Comment by Nyholm** [comment]
I agree with Mark. 

* It is nothing in this PR that is Guzzle specific. 
* There is an excellent package (https://github.com/zendframework/zend-httphandlerrunner) that already do this and only this. 
* I dont see any benefits of putting this feature in this library. 

Im 👎

**Comment by plashenkov** [comment]
> seems to be a direct copy of the Zend SAPI Emitter

No, that's not a "direct copy". Just to clarify. Though I looked at it as well as other packages implementing such functionality. Structure is almost the same because it's hard to invent something better.

Well, OK. I understand.

**Pull request closed without merging**